### PR TITLE
Implement Crazyhouse

### PR DESCRIPTION
**Pull request by ianfab**
All crazyhouse rules should be implemented and perft results seem to be correct.

Drops are printed as piece@square, e.g., `P@e5`. In the move encoding 5 additional bits are used for drops: 1 for the move type DROP, and 4 for the dropped piece.

Known issues:
- so far hash collisions are likely, because the pieces in hand are not taken account of in the calculation of hash keys.
- movepicking should probably be improved
- evaluation does not take account of pieces in hand

In summary, except for the rules nothing has been adapted for crazyhouse yet. Further validation of the move generation should be performed, especially to test for the handling of special rules like the capturing of promoted pieces and subsequent drops.

### CODE REVIEW

**Comment by ddugovic** [comment]
Excellent, I'll start writing front end code (lichess) to facilitate user testing...

**Comment by ianfab** [comment]
Surprisingly, Stockfish already is quite decent at playing crazyhouse. E.g. in the position `r1b2rk1/1pp1qp2/5n1p/1Bp5/4P3/5P2/PP3P1P/R2Q1RK1[NPPPBPpnbn] b - - 31 16` it finds the correct mate in 4 involving four piece drops at depth 4. However, in the exact same position a segmentation fault occurs at depth 16.

**Comment by ddugovic** [comment]
When I compile with `debug=yes` I observe:

```
info string variant crazyhouse startpos r1b2rk1/1pp1qp2/5n1p/1Bp5/4P3/5P2/PP3P1P/R2Q1RK1[BNPPPPbnnp] b - - 31 16

stockfish: position.cpp:1400: void Position::undo_move(Move): Assertion `empty(from) || type_of(m) == CASTLING' failed.

Thread 2 "stockfish" received signal SIGABRT, Aborted.
[Switching to Thread 0x7ffff6b43700 (LWP 21684)]
0x00007ffff6f85418 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
54      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff6f85418 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff6f8701a in __GI_abort () at abort.c:89
#2  0x00007ffff6f7dbd7 in __assert_fail_base (fmt=<optimized out>, assertion=assertion@entry=0x45f1a0 "empty(from) || type_of(m) == CASTLING", 
    file=file@entry=0x45f30a "position.cpp", line=line@entry=1400, 
    function=function@entry=0x45f510 <Position::undo_move(Move)::__PRETTY_FUNCTION__> "void Position::undo_move(Move)") at assert.c:92
#3  0x00007ffff6f7dc82 in __GI___assert_fail (assertion=assertion@entry=0x45f1a0 "empty(from) || type_of(m) == CASTLING", 
    file=file@entry=0x45f30a "position.cpp", line=line@entry=1400, 
    function=function@entry=0x45f510 <Position::undo_move(Move)::__PRETTY_FUNCTION__> "void Position::undo_move(Move)") at assert.c:101
#4  0x000000000043437b in Position::undo_move (this=this@entry=0x7ffff6b44130, m=m@entry=(MOVE_NULL | unknown: 655406)) at position.cpp:1400
#5  0x0000000000438714 in (anonymous namespace)::qsearch<(<unnamed>::NodeType)0u, false>(Position &, Search::Stack *, Value, Value, Depth) (
    pos=..., ss=0x7ffff6b41580, alpha=-449, beta=-448, depth=DEPTH_ZERO) at search.cpp:1596
#6  0x000000000043d227 in (anonymous namespace)::search<(<unnamed>::NodeType)1u>(Position &, Search::Stack *, Value, Value, Depth, bool) (pos=..., 
    ss=0x7ffff6b41550, alpha=448, alpha@entry=-32001, beta=beta@entry=VALUE_INFINITE, depth=<optimized out>, cutNode=cutNode@entry=false)
    at search.cpp:1187
#7  0x000000000043f1c5 in Thread::search (this=this@entry=0x7ffff6b44010) at search.cpp:482
#8  0x0000000000441538 in MainThread::search (this=0x7ffff6b44010) at search.cpp:350
#9  0x0000000000444131 in Thread::idle_loop (this=0x7ffff6b44010) at thread.cpp:114
#10 0x00007ffff78f0c80 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#11 0x00007ffff7bc16fa in start_thread (arg=0x7ffff6b43700) at pthread_create.c:333
#12 0x00007ffff7056b5d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

**Comment by ianfab** [comment]
Thanks. I deliberately set the from_sq of piece drops to SQ_NONE to provoke such warnings. My latest PR fixes a few of those issues, but apparently not all.
